src/app/services/configuration-control.service.ts

```diff
@@ -0,0 +1,517 @@
+import { Injectable } from '@angular/core';
+import { FormControl, FormGroup, Validators } from '@angular/forms';
+import { AclService } from './acl.service';
+import { orderBy, map, cloneDeep, reduce, assign } from 'lodash';
+
+export class ConfigurationBase<T> {
+  value: T | undefined;
+  key: string;
+  label: string;
+  description: string;
+  required: boolean;
+  readonly: string;
+  editable: boolean;
+  mandatory: string;
+  maximum: string;
+  minimum: string;
+  order: number;
+  length: string; // hold password max length
+  controlType: string;
+  type: string;
+  options: { key: string, value: string }[];
+  editorOptions: {};
+  fileName: string;
+  file: File;
+  validFileExtension?: boolean;
+  validity?: string;
+  validityExpression?: string;
+
+  constructor(options: {
+    value?: T;
+    key?: string;
+    label?: string;
+    description?: string,
+    required?: boolean;
+    readonly?: string;
+    editable?: boolean,
+    mandatory?: string,
+    maximum?: string,
+    minimum?: string;
+    order?: number;
+    controlType?: string;
+    type?: string;
+    options?: [];
+    length?: string;
+    editorOptions?: {};
+    file?: string;
+    files?: File;
+    validFileExtension?: boolean;
+    validity?: string;
```

Comment:
rule?
deprecated?

Reply:
We can add but not sure if those are available in FogLAMP yet.I haven't seen.

Reply:
```
_optional_items = sorted(['readonly', 'order', 'length', 'maximum', 'minimum', 'rule', 'deprecated', 'displayName',
                          'validity', 'mandatory', 'group'])
```